Sources/Basics/Terminator.swift

```diff
@@ -0,0 +1,112 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2022 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+ */
+
+import Dispatch
+import Foundation
+import TSCBasic
+
+public typealias TerminationHandler = () throws -> Void
+
+#warning("make observabilityScope non optional")
+public class Terminator {
+    public typealias RegistrationKey = String
+
+    private let killTimeout: DispatchTimeInterval
+    private let observabilityScope: ObservabilityScope?
+
+    private let registry = ThreadSafeKeyValueStore<String, (name: String, handler: TerminationHandler)>()
+    private let terminationQueue = DispatchQueue(label: "org.swift.swiftpm.terminator")
+
+    private let terminated = ThreadSafeBox<Bool>(false)
+
+    public init(
+        killTimeout: DispatchTimeInterval = .seconds(30),
+        observabilityScope: ObservabilityScope?
+    ) {
+        self.killTimeout = killTimeout
+        self.observabilityScope = observabilityScope
+    }
+
+    @discardableResult
+    public func register(name: String, handler: @escaping TerminationHandler) -> RegistrationKey {
+        if self.terminated.get() ?? false {
+            self.observabilityScope?.emit(debug: "not registering '\(name)' with terminator as its already cancelled")
+            // TODO: do something better in this case
+            return "noop" // nothing was done
+        }
+        let key = UUID().uuidString
+        self.observabilityScope?.emit(debug: "registering '\(name)' with terminator")
+        self.registry[key] = (name: name, handler: handler)
+        return key
+    }
+
+    public func register(_ process: TSCBasic.Process) -> RegistrationKey {
+        self.register(name: "\(process.arguments.joined(separator: " "))", handler: {
```

Comment:
Would it make sense to require the process to have been launched already and use a key that there can't be two at once of, like the pid?  We can probably make things like plugin command lines unique but there's no guarantee that they are (since plugins communicate with the host more interactively than this).

Either way it's a details, but the part that affects this is whether to require the process to already be running.

Reply:
the name is not the unique key, its only for presentation

Reply:
Oh right — I think I saw that later down in the PR.